### DISCUSSION TITLE
helm chart for flux2

### DISCUSSION THREAD

**Discussion by nab-gha** [fluxcd/flux2#431] [Proposals] [answered]
Some customers exclusively use helm to deploy applications to Kubernetes. They will want be able to install gotk using helm.

At the customer I'm working with I manually created a helm chart that containing source-controller, helm-controller and the Kraan controller that uses it, see https://github.com/fidelity/kraan/tree/master/chart. I added facility to provide http proxy in the chart. This approach involves some manual steps when upgrading the source-controller and helm-controller in the Kraan application.

Maybe we could add functionality to the toolkit installer to generate helm charts so that a public helm chart for the gotk components can be maintained?

**Comment by hiddeco** [comment]
At this moment we do not plan to maintain a Helm chart in the foreseeable future. 

Helm charts tend to be maintenance (and support) heavy, and in combination with the limitations around rolling out Custom Resource Definitions it is not a perfect fit for any of the GitOps Toolkit components at present. Introducing a way to generate the Hem chart would not reduce this, and will still result in people wanting to add features to the chart generator, as proven by your:

> I added facility to provide http proxy in the chart. This approach involves some manual steps when upgrading the source-controller and helm-controller in the Kraan application.

I strongly believe it should be possible for the users that really need it to collaborate on a "unofficial" Helm chart until we may have an official chart one day. That is the magic of OSS. This means the maintainers can focus on things that serve all users, as there are countless things that still need to be done.

**Reply by nab-gha** [reply]
@hiddeco I agree that this is lower priority than adding functionality to gotk.

The problem with an "unofficial" approach is that it quickly becomes out of date as gotk evolves.

Would the maintainers of gotk be open to a community contribution to the installer to generate helm charts?

**Reply by hiddeco** [reply]
That would circle back to

> Introducing a way to generate the Hem chart would not reduce this, and will still result in people wanting to add features to the chart generator, [...]

for which I am not really in favor. But this has not been debated so the view of the others may be different.

**Reply by scottrigby** [reply]
A big part of the benefit of Helm's [post-renderer](https://helm.sh/docs/topics/advanced/#post-rendering) feature is allowing users to add their use cases without every imaginable use case having to be added to a chart. I'll check in with the team about this, as this approach significantly reduces chart maintenance.

**Reply by squaremo** [reply]
My experience from Flux's Helm chart is that the text templates pretty quickly become difficult to maintain, and the values specification a giant cruftball. The only remedy for this I can come up with is to make your programs have (next to) nothing configurable, and stand firm on adding non-functional embellishments.

I don't think adding `kustomize` to the mix makes the situation better -- it expands the contract of the chart to its entire output, since any change could break a patch somewhere.

**Reply by scottrigby** [reply]
Agree, charts need to be well maintained to stay simple. Luckily we have someone on our team with lots of experience maintaining Helm chart source and repos 😄

**Comment by scottrigby** [comment]
related: link to discussion on this from last Flux meeting (Nov 5th), see [last bullet item in notes](https://docs.google.com/document/d/1l_M0om0qUEN_NNiGgpqJ2tvsF2iioHkaARDeh6b70B0/edit#heading=h.4w6ysomi4xa).

**Comment by scottrigby** [comment]
Another fun note. We currently also have other ways of installing Flux than the CLI, for example the Terraform provider. And in dev mode, you may also install Flux with just kubectl (then create git repo history on cluster, then place file you created with kubectl in your git repo), the same steps that would be required with a Flux 2 Helm chart.

**Reply by hiddeco** [reply]
The Terraform provider re-uses part of the Flux CLI, as it makes use of the [`manifestgen` packages](https://github.com/fluxcd/flux2/tree/main/pkg/manifestgen) to generate the manifests. It requires thus much less maintenance as it builds upon already existing bits and pieces: `Terraform module <- manifestgen <- Kustomize bases`.

The same applies to the `kubectl` part, as we offer a simple generated `install.yaml` with every release (see attachments on https://github.com/fluxcd/flux2/releases), that is again, generated from the Kustomize bases.

**Reply by dmccaffery** [reply]
A few comments:

1. There are some serious flaws in the current kustomizations:
  * ClusterRole / ClusterRoleBinding is used regardless of whether or not watch-all-namespaces is true
  * ClusterRoleBinding is associated with the default service account (bad ju-ju)
  * The name of cluster-wide resources is hardcoded and not configurable (without many handcrafted patches) -- which means that multiple installations into different namespaces will step on each other and could break each other if the versions do not remain synchronised
2. There are also some serious flaws with the design of the terraform provider -- mostly as a result of 1 above
  * the gavinbunney/kubectl provider that is recommended is naive and doesn't ensure ownership of resources contrary to the design intent of terraform -- multiple applies from different plans can overwrite the same resources (even in the same namespace) -- or resources not created with terraform in the first place
  * this conflates terraform design with kubernetes design in horrible ways that are inconsistent:
  
![image](https://user-images.githubusercontent.com/1803684/112042435-67773c00-8b3f-11eb-8f19-23555d38773f.png)

3. Agree that any implementation of a helm chart *SHOULD NOT* exceed the capabilities of the kustomize generation and baseline requirements of the system, except to implement best practices around labels, annotations, and naming conventions of resources as the kustomization should be doing anyway -- any extension of the default capabilities should be achieved via parent charts owned / maintained by the customer and/or post-renderer.
4. Helm, at its core, is just go templates -- it doesn't have to be any more complicated than the go templates that are already defined in the src: https://github.com/fluxcd/flux2/blob/main/pkg/manifestgen/install/templates.go#L27 -- I would argue that kustomize is considerably more difficult to write and maintain -- the patches construct is fine for extremely simple scenarios, but falls over with even the slightest hint of complexity (such as a feature flag) -- lets hope the watch-all-namespaces argument stays at index 0 forever, for instance: https://github.com/fluxcd/flux2/blob/bd41406aaa13c3c04e6342a5c596362fcade09e9/pkg/manifestgen/install/templates.go#L65

As of now, the installation and maintenance mechanisms provided for flux2 do not meet the needs of any large-scale deployment. We manage clusters with thousands of pods and over a hundred engineers. These are not multi-tenant clusters, either. They are very large product-specific clusters.

**Reply by cwrau** [reply]
You really took the words out of my mouth with number 4, that's why I really don't get the hype around kustomize.

Sure, if you just need to change the names of the resources and have common labels then it's perfect, but anything even slightly more complex is way out of scope

And as you pointed out with the reference, even when using kustomize, people still use templates, so why not use Helm immediately? That way, it'd be just one standard tool instead of a standard tool and non-standard workarounds around that tool

**Comment by dmccaffery** [comment]
To move forward, I'm happy to contribute to a community-driven helm chart -- the question is where this chart should live. Would a new organisation called "fluxcd-community" be appropriate (similar to prometheus-community)?

**Reply by dmccaffery** [reply]
@hiddeco / @stefanprodan (hate to tag you, but just want to know if fluxcd-community is okay by you) -- I couldn't join the community meeting last Thursday due to overlaps :(

**Reply by hiddeco** [reply]
@dmccaffery I just created the `fluxcd-contribx` organization, as the `prometheus-community` is governed by the Prometheus organization as well (and we probably want to "recognize" other community driven projects at a later moment, for example the OpenShift support efforts).

What would you like the repository to be called?

**Reply by dmccaffery** [reply]
I was suggesting "fluxcd-community", but it doesn't matter to me :)

For now "helm-charts" is a good starting point for the repo name. I was going to go down the avenue of forking all the repos and creating a chart per component as I was consuming the existing kustomize manifests to help generate the charts (and support an eventual merge back), but I can rethink that strategy if that's not appropriate.

**Reply by makkes** [reply]
@dmccaffery I'm happy to help on your effort to come up with a community-maintained chart. We are currently maintaining our own sort of chart (actually as part of a bigger chart) internally but would also favor a public one. Making sure that e.g. proper tests are in place (esp. for upgrades etc.) can be tedious from my experience. 🙂

**Reply by dmccaffery** [reply]
The more the merrier! I won't have an opportunity to pull in what I've done already until the weekend; it's a busy week for me -- but happy to collaborate.

Example of the strategy I was going for is here:

https://github.com/fluxcd/source-controller/pull/314/commits/71ea919996f4407ccefffd3ba16fbc4f45a14ecf

**Reply by stefanprodan** [reply]
I would not copy/paste thousands of lines of YAMLs from all flux repos manually to some other org. @dmccaffery your PR raises a lot of questions for me, first of all the manifests in each repo are for dev only, the actual manifests that are subject to e2e testing are generated with [flux2/pkg/manifestgen](https://github.com/fluxcd/flux2/tree/main/pkg/manifestgen). The controllers are not guaranteed to work together unless they are part of a flux2 release, the chart must enforce this version matching between APIs, container images and CLI. Also you can't use Helm to upgrade CRDs if you place them in the `crds` dir, we need to provide a dedicated chart for each API separate to the controller.

**Reply by dmccaffery** [reply]
Hmm. If what you say is true and the versions of the controllers, cli, and crds must all exactly align across all controllers in order to function; then this is a distributed monolith.

I'll rethink the strategy, then. My assumption was that each controller operated independently and that any api between them would limit breaking changes. In other words, it should be safe to upgrade one controller independently of the others as long as it's not versioned as a breaking change.

I don't understand why these are broken down into poly repos if they are tightly coupled in this way. Regardless, I will rethink the strategy. I was purposefully avoiding tight coupling under a false assumption that this was the design intent.

**Reply by stefanprodan** [reply]
> it should be safe to upgrade one controller independently of the others as long as it's not versioned as a breaking change

In theory this is true, but as all reconciliation APIs are depending on the source ones this could break if it's not validated by e2e.

**Reply by hiddeco** [reply]
> Hmm. If what you say is true and the versions of the controllers, cli, and crds must all exactly align across all controllers in order to function; then this is a distributed monolith.
> I don't understand why these are broken down into poly repos if they are tightly coupled in this way.

They are all depending on unstable APIs that are still under development, this means things may be added to a controller another controller depends on, and from that moment having a version equal or higher to when a feature was introduced becomes a requirement to support all features.

This does not make it a monolith, it simply means the services are all still under active development. Once they reach `1.0.0`, this problem goes away.

**Reply by makkes** [reply]
We are currently using `flux install --export` and doing some transformations on the results. Maybe that could serve as an automation start?
```
.PHONY: flux-update
.ONESHELL:
flux-update: $(FLUX_BIN)
	tmpdir=$(shell mktemp -d)
	curdir=$(shell pwd)
	cd $${tmpdir}
	$${curdir}/$(FLUX_BIN) install -n flux-system --export > flux.yaml
	$${curdir}/$(KUBERNETES_SPLIT_YAML_BIN) --outdir=./flux --tpl_flat "{{.X.Outdir}}/flux-{{.X.ShortKind}}_{{.Metadata.Name}}.yaml" flux.yaml
	rm flux/flux-namespace_flux-system.yaml
	find flux -name '*.yaml' -not -name 'flux-crd_*.yaml' -print0 | xargs -0 sed -i 's/flux-system/{{ .Release.Namespace }}/'
	chmod 0664 ./flux/flux-*.yaml
	rm -rf $${curdir}/chart/templates/flux/* $${curdir}/chart/crds/flux/*
	mv ./flux/flux-crd* $${curdir}/chart/crds/flux/
	mv ./flux/flux-*.yaml $${curdir}/chart/templates/flux/
	cd $${curdir}
	rm -rf $${tmpdir}
```

**Reply by pierluigilenoci** [reply]
In case there is a need, I am also available to participate in the maintenance of the chart.

**Comment by mewzherder** [comment]
@hiddeco sorry to come in late, but since @alisondy @dholbach @staceypotter and I are working to build out Contributor Experience for Flux, this naming will create a lot of confusion. Can we support the suggestions for fluxcd-community instead? If I understand the repo correctly, these are community contributions, and not about building out a positive contributor experience (which is what we commonly call ContribuX). Note that this is coming from a discussion that we're having amongst us w/ @scottrigby and others. Thanks for the consideration. ;)

**Comment by pierluigilenoci** [comment]
Is there any news about this?
In case of need, I am also available to maintain the chart.

**Comment by haarchri** [comment]
@stefanprodan  can you create the contrib repo ? afterwards i am also available to maintain the chart with the others together - feedback is welcome

**Reply by hiddeco** [reply]
Created one awhile ago and gave someone here access, but I think it never become something. I can invite you to this repository if you would like.

**Reply by hiddeco** [reply]
https://github.com/fluxcd-community/helm-charts

**Reply by pierluigilenoci** [reply]
@haarchri @hiddeco if you need help I'm available.

**Reply by haarchri** [reply]
@hiddeco i would say that @pierluigilenoci  and i can start working on that - @pierluigilenoci  lets have a session in slack o.a.

**Reply by hiddeco** [reply]
You should both have an invite, if you run into anything, I can be reached in #flux-contributors (`@hidde`) on CNCF Slack.

**Reply by pierluigilenoci** [reply]
@hiddeco thank you

**Reply by dmccaffery** [reply]
I've been meaning to merge mine but have been pulled into other things. Now that the dust has settled around flux v2, I'm happy to bring in what I have as a starting point.

**Reply by pierluigilenoci** [reply]
thanks @dmccaffery

**Reply by haarchri** [reply]
thanks - my idea is to merge first an very basic helm-chart like the following https://github.com/haarchri/helm-charts - and then we can start to work for generator and more flexible configuration - what you guys think ? if we would go i can create the PR in community repo

**Reply by pierluigilenoci** [reply]
I'm ok with that.

**Comment by haarchri** [comment, marked as answer]
we released a very basic pre-alpha 0.0.1 version available here: https://github.com/fluxcd-community/helm-charts - for all feel to open issues & create PRs

**Comment by pierluigilenoci** [comment]
@paulcarlton-ww @haarchri @stefanprodan @hiddeco at this point I believe that this issue can be closed.

**Comment by elliottpope** [comment]
Can this discussion be labeled with the new `area/install` label since it took me a while to find this and I was specifically looking at why the Helm chart is not officially supported to inform whether or not I use Terraform or Helm to install Flux v2
